# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Крицын Данила, вариант 37.

## Цель работы

Сформировать представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

Аддитивной цепочкой для $ n \in ℕ $ называется последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-то двух предыдущих:
$$ a_i = a_j + a_k, \quad k \le j \le i, \quad i = 1, 2, ..., r $$

$l(n) = r$ - наименьшая длина аддитивной цепочки для $ n \in ℕ $.

Для метода наименьших множителей: $ \quad l(mn) \le l(m) + l(n) $

Для m-арного метода: $ \quad l(n) \le m - 2 + (k + 1)t \quad [m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j}] $

Для SX-метода: $ \quad l(n) \le \lambda (n) + \nu (n) - 1 $


### Свойства аддитивных цепочек:
* Полагается строгое возрастание элементов цепочки:
$ 1 = a_0 < a_1 < a_2 < ... < a_n = n $
* Одинаковые числа в цепочке можно опустить
* Пара $ (j, k), 0 \le k \le j < i $ называется шагом $i$
* Если $\exists$ более чем 1 пара $ (j, k) $, полагаем $ max \hspace{0.2cm} j $

### Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $ (линейный шаг)
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $, где $ \lambda (n) = \lfloor lb(n) \rfloor$

### Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым

### Теорема:

Если аддитивная цепочка содержит $d$ и $f = r - d$ неудвоений, то $n \le 2^{d-1} F_{f+3}$, где $F_j$ - число Фиббоначи

### Следствие:

Если аддитивная цепочка содержит $f$ удвоений и $S$ малых шагов, то

${S \le f \le} {S \over {1 - lb(\varphi)}}$, где $\varphi = {{\sqrt{5} + 1} \over 2} $ - золотое сечение

### Алгоритм Брауэра:

Для $n \in ℕ$ при заданном $k \in ℕ$ можно построить цепочку Брауэра с помощью рекуррентной формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Данная цепочка имеет длину
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимализирована для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

**Ход алгоритма:**

* Задаётся некий параметр $k$ для $n$.
Выполняется вычисление вспомогательных чисел:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d => n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d => q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d,$ следовательно $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ..., 2^k - 1, \\
2q_s, 4q_s, 8q_s, ..., 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ..., 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
..., \\
..., 2^k q_1, 2^k q_1 + r_1 = n.$$

### Алгоритм Яо:

* Обладает такой же вычислительной мощностью, что и алгоритм Брауэра
* Начало похоже: $k \ge 2$, $n$ раскладывается в $2^k$-ой системе счисления:
$$ n = \sum_{i = 0}^j a_i 2^{ik} , a_j \ne 0 $$
* Введём функцию d:
$$ d(z) = \sum_{i: a_i = z} 2^{ik} $$

**Ход алгоритма:**

* Базовая последовательность: $1, 2, 4, ..., 2^{\lambda(n)}$
* Вычисление значений $d(z)$ для всех $z \in \{ 1, 2, ..., 2^k - 1\}, \quad d(z) \ne 0$
* Вычисление $zd(z)$ для всех $z$
* n раскладывается в виде
$$ n = \sum_{z = 1}^{2^k - 1} zd(z) $$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность вычислений бинарным методом и методом множителей для $x^n$ для 2-3 значений $n$ ($n \ge 30$). Сравнить количество операций для каждого метода, сделать выводы.

**n = 80:**<br>
Двоичный метод:<br>

N | Y | Z
--- | --- | ---
80 | 1 | x
40 | 1 | x^2
20 | 1 | x^4
10 | 1 | x^8
5 | 1 | x^16
2 | x^16 | x^32
1 | x^16 | x^4
0 | x^80 | x^128

Метод множителей:<br>

$$ 80 = pq = 2 \cdot 40 $$
$$ y = x^2\\
   y^{40} = (y^8)^5 = (((y^2)^2)^2)^5 = ((((y^2)^2)^2)^2)^2 \cdot (((y^2)^2)^2)
$$
   
**n = 126:**<br>
Двоичный метод:<br>

N | Y | Z
--- | --- | ---
126 | 1 | x
63 | 1 | x^2
31 | x^2 | x^4
15 | x^6 | x^8
7 | x^14 | x^16
3 | x^30 | x^32
1 | x^62 | x^64
0 | x^126 | x^128

Метод множителей:<br>

$$ 126 = pq = 2 \cdot 63 $$
$$ y = x^2\
   y^{63} = y^{62} \cdot y = (y^2)^{31} \cdot y \\
   z^{31} = z^{30} \cdot z = (z^5)^6 \cdot z = ((z^2)^2 \cdot z)^6 \cdot z = (((z^2)^2 \cdot z) ^ 2 \cdot ((z^2)^2 \cdot z)) ^ 2 \cdot z\\
   y^{63} = y^{62} \cdot y = (y^2)^{31} \cdot y = ((((y^2)^2 \cdot y) ^ 2 \cdot ((y^2)^2 \cdot y)) ^ 2 \cdot y) ^ 2 \cdot y\\
   x^2 = (((((x^2)^2 \cdot x) ^ 2 \cdot ((x^2)^2 \cdot x)) ^ 2 \cdot x) ^ 2 \cdot x) ^ 2
$$

Количество операций:<br>

n | двоичный метод | метод множителей
--- | --- | ---
80 | 9 | 10
126 | 13 | 13

Видно, что для чисел с небольшим количеством единиц в двоичной записи двоичный метод работает практически за одну операцию меньше, чем метод множителей (было проверено также для $ n = 98, 69 $). Для чисел с большим количество единиц в двоичной записи двоичный метод завершает свою работу примерно за столько же операций, за сколько завершается метод множителей. Скорее всего, метод множителей работает лучше на числах, для которых $p > 2$, а время работы двоичного метода близко к максимальному ($p$ близко к $2^k - 1$ по количеству единиц в двоичной записи).

2. Реализовать алгоритм Брауэра для вычисления приближённых аддитивных цепочек для различных чисел при варьировании параметра k, сопоставить длины полученных аддитивных цепочек с минимальной аддитивной цепочкой для заданного числа. Сделать выводы.

In [14]:
import random

def _lambda(n):
    l = 0
    while n > 0:
        n = int(n / 2)
        l += 1
    return l - 1

def min_chain_length(n):
    return int(_lambda(n) + _lambda(n) / _lambda(_lambda(n)) + (_lambda(n) * _lambda(_lambda(_lambda(n)))) / (_lambda(_lambda(n)) ** 2))


def chain_remove_dups(chain):
    ret = list(dict.fromkeys(chain))
    if ret[-1] == 0:
        del ret[-1]
    return ret

def chain_to_str(chain):
    return " ".join(map(str, chain[:min(len(chain), 16)])) + (" ..." if len(chain) > 16 else "")


def brouwer(n, k):
    if type(n) != int and type(n) != Integer:
        raise(TypeError("n should be integer"))
    if type(k) != int and type(k) != Integer:
        raise(TypeError("k should be integer"))

    d = 2 ** k

    rarr = [] # array of "remainders" (d-ary number digits)
    q = int(n / d)
    rarr.append(n % d)
    sqmax = rarr[0] # maximum number in the digits of n

    i = 0
    while q >= d:
        rarr.append(q % d)
        q = int(q / d)
        if rarr[i] > sqmax:
            sqmax = rarr[i]
        i += 1
    rarr.append(q)
    sqmax = max(q, sqmax)

    chain = [0] * (sqmax + (len(rarr) - 1) * (k + 1)) # addition chain
    for i in range(1, sqmax + 1):
        chain[i-1] = i # starting sequence 1, 2, 3, ..., b, where b is maximum number in the digits of n

    chain[sqmax - 1] = q
    for i in range(0, len(rarr) - 1): # i - digit number
        for j in range(0, k): # j - number to double
            chain[sqmax + i * (k + 1) + j] = chain[sqmax + i * (k + 1) + j - 1] * 2
        chain[sqmax + i * (k + 1) + k] = chain[sqmax + i * (k + 1) + k - 1] + rarr[len(rarr) - 2 - i]
    chain[sqmax - 1] = sqmax

    return chain_remove_dups(chain)


random.seed(641202498)
n_set = []
for i in range(0, 5):
    n_set.append(random.randint(17, 127))
for i in range(0, 5):
    n_set.append(random.randint(127, 65535))

for k in [1, 2, 3, 6, 9, 16]:
    print("k = ", k, ":", sep = "")
    for n in n_set:
        chain = brouwer(n, k)
        print("\t n = {:6} l_B(n) = {:6} l(n) = {:6} | {}".format(n, len(chain), min_chain_length(n), chain_to_str(chain)))

k = 1:
	 n =     85 l_B(n) =     10 l(n) =     10 | 1 2 4 5 10 20 21 42 84 85
	 n =     59 l_B(n) =     10 l(n) =      8 | 1 2 3 6 7 14 28 29 58 59
	 n =     21 l_B(n) =      7 l(n) =      7 | 1 2 4 5 10 20 21
	 n =     24 l_B(n) =      6 l(n) =      7 | 1 2 3 6 12 24
	 n =    117 l_B(n) =     11 l(n) =     10 | 1 2 3 6 7 14 28 29 58 116 117
	 n =  61900 l_B(n) =     24 l(n) =     21 | 1 2 3 6 7 14 15 30 60 120 240 241 482 483 966 967 ...
	 n =  22175 l_B(n) =     24 l(n) =     20 | 1 2 4 5 10 20 21 42 43 86 172 173 346 692 1384 1385 ...
	 n =  62160 l_B(n) =     23 l(n) =     21 | 1 2 3 6 7 14 15 30 60 120 121 242 484 485 970 971 ...
	 n =   1200 l_B(n) =     14 l(n) =     14 | 1 2 4 8 9 18 36 37 74 75 150 300 600 1200
	 n =  51251 l_B(n) =     22 l(n) =     21 | 1 2 3 6 12 24 25 50 100 200 400 800 1600 1601 3202 3203 ...
k = 2:
	 n =     85 l_B(n) =     10 l(n) =     10 | 1 2 4 5 10 20 21 42 84 85
	 n =     59 l_B(n) =      9 l(n) =      8 | 1 2 3 6 12 14 28 56 59
	 n =     21 l_B(n)

Видно, что оптимальное $k(n)$ совсем невелико по сравнению с $n$, более точно оно определяется как
$$ k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n) $$ для достаточно больших $n$.<br>
В данном случае $n$ были недостаточно большими: оптимальное $k$ не превышает 1, при этом на практике оно приближённо равно 2: видно, что и для первых пяти чисел в диапазоне $[17, 127]$, и для последних пяти чисел в диапазоне $[127, 65535]$ $l_B(n)$ только увеличивается при увеличении $k \ge 2$.<br>
Кроме того, слишком большое $k$ ведёт к деградированию цепочек, построенных алгоритмом Брауэра, до рядов натуральных чисел. Это можно проследить в первую очередь на меньших числах: при $k = 6$ это $n = 21, 24, 59$, при $k = 9$ это все выбранные числа $n = 21, 24, 59, 85, 117$.<br>
Очевидно, полное совпадение цепочки метода Брауэра с рядом натуральных чисел происходит при таком $k$, что $2^k \ge n$. Это можно увидеть при $k = 16$, ведь $2^{16} = 65536$, что больше любого выбранного $n$.<br>
Однако, увеличение длины цепочки для чисел из второго диапазона $[127, 65535]$ уже заметно и при $k = 9$: например, для $n = 61900$ $l_B(n) = 469$, в то время как минимальная $l(n) = 21$.<br>

Таким образом, чтобы алгоритм Брауэра строил аддитивные цепочки адекватной длины, $k$ должно быть куда меньше $n$ и для достаточно больших $n$ вычисляться по формуле $ k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n) $, а для небольших $n$ быть положено равным $1$ или $2$ (разница должна быть не очень заметной).

3. Реализовать алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки для заданного числа. Протестировать алгоритм минимум для 5 значений n > 1000. Указать, сколько времени потребовалось на поиск цепочки и какая цепочка получилась. Сравнить с предыдущими методами, сделать выводы.

In [15]:
import math
import random

def _lambda(n):
    l = 0
    while n > 0:
        n = int(n / 2)
        l += 1
    return l - 1

def _nu(n):
    l = 0
    while n > 0:
        if n % 2 != 0:
            l += 1
        n = int(n / 2)
    return l


def min_chain_length(n):
    return int(_lambda(n) + 1)

def chain_from_indvec(indvec):
    indvec_len = len(indvec)
    nums = [0] * (indvec_len + 1)
    nums[0] = 1

    for i in range(0, indvec_len):
        nums[i + 1] = nums[i] + nums[indvec[i] - 1]
    return nums

def chain_to_str(chain):
    return " ".join(map(str, chain[:min(len(chain), 40)])) + (" ..." if len(chain) > 40 else "")


def indvec_gen(n, start = 1):
    indvec = [0] * n
    for i in range(0, n):
        indvec[i] = i + start
    return indvec

def indvec_is_min(indvec):
    for i in indvec:
        if i > 1:
            return False
    return True

def indvec_dec(indvec):
    indvec_len = len(indvec)

    for i in range(0, indvec_len):
        if indvec[indvec_len - 1 - i] > 1:
            indvec[len(indvec) - 1 - i] -= 1
            for j in range(indvec_len - i, indvec_len):
                indvec[j] = j + 1
            return

def indvec_get_last_num(indvec):
    indvec_len = len(indvec)
    nums = [0] * (indvec_len + 1)
    nums[0] = 1

    for i in range(0, indvec_len):
        nums[i + 1] = nums[i] + nums[indvec[i] - 1]
    return nums[indvec_len]


def indvec_crush(n):
    tries = [1, 1]

    lmin = math.ceil(math.log(n, 2))
    lmax = _lambda(n) + _nu(n) - 1

    for m in range(lmin, lmax + 1):
        q = int(m / 2)
        iv_fix = indvec_gen(q)

        while True: # processing all possible fixed parts
            iv_mut = indvec_gen(m - q, q + 1)

            iv_fix.append(iv_mut[0])
            aq = indvec_get_last_num(iv_fix)
            del iv_fix[-1]
            amin = 1 + m - q          # a_{q+1} + m - q
            amax = aq * 2 ** (m - q)  # a_{q+1} * 2^{m-q}

            if n > amax or n < amin:  # n is not in [a_min, a_max], skipping this fixed part
                if indvec_is_min(iv_fix):
                    break
                indvec_dec(iv_fix)
                continue

            while True: # processing all possible mutable parts
                if indvec_get_last_num(iv_fix + iv_mut) == n:
                    return (iv_fix + iv_mut, tries)

                if indvec_is_min(iv_mut):
                    break
                indvec_dec(iv_mut)
                tries[1] += 1

            if indvec_is_min(iv_fix):
                break
            indvec_dec(iv_fix)
            tries[0] += 1


random.seed(295316871)
n_set = []
for i in range(0, 5):
    n_set.append(random.randint(1001, 1250))
print("{}\n".format(n_set))
for n in n_set:
    res = indvec_crush(n)
    print("n = {:4}   l*(n) = {:4}   l(n) = {:4}   fix. parts tried: {:6}   mut. parts tried: {:6} \n{}\n".format(n, len(res[0]), min_chain_length(n), res[1][0], res[1][1], chain_to_str(chain_from_indvec(res[0]))))

[1138, 1086, 1008, 1142, 1237]

n = 1138   l*(n) =   13   l(n) =   11   fix. parts tried:    825   mut. parts tried: 6772538 
1 2 4 8 16 32 34 68 136 272 544 1088 1122 1138

n = 1086   l*(n) =   13   l(n) =   11   fix. parts tried:    943   mut. parts tried: 13498407 
1 2 4 6 12 24 30 60 120 240 480 960 1080 1086

n = 1008   l*(n) =   12   l(n) =   10   fix. parts tried:    152   mut. parts tried: 898256 
1 2 4 8 16 24 28 56 112 224 448 896 1008

n = 1142   l*(n) =   14   l(n) =   11   fix. parts tried:   1545   mut. parts tried: 47912309 
1 2 4 8 16 32 64 66 132 264 528 1056 1122 1138 1142

n = 1237   l*(n) =   14   l(n) =   11   fix. parts tried:   1541   mut. parts tried: 49159794 
1 2 4 8 16 32 36 37 74 148 296 592 1184 1221 1237



По результатам видно, что чем ближе число к степени двойки (начиная с младших разрядов), тем меньше необходимо перебрать векторов индексов. Например, для числа с наименьшим количеством перебранных как фиксированных, так и изменяемых частей $n = 1008_{10} = 11 1111 0000_2$. Для числа с наибольшим количеством перебранных векторов индексов $n = 1142_{10} = 100 0111 0110$, что довольно далеко от степени двойки и тем более имеет нули во 2, 3 и 4 старших разрядах.

По сравнению с методом Брауэра отношение длин цепочек к числам получаются примерно одинаковыми, и разумеется большими чем минимальные длины цепочек $l(n) = \lceil lb(n) \rceil$ .

Также видно, что алгоритм имеет среднюю сложность $O(\sum(m-1)!)$, где $m$ - длина минимальной звёздной цепочки.

In [2]:
import math
import random

def _lambda(n):
    l = 0
    while n > 0:
        n = int(n / 2)
        l += 1
    return l - 1

def _nu(n):
    l = 0
    while n > 0:
        if n % 2 != 0:
            l += 1
        n = int(n / 2)
    return l


def min_chain_length(n):
    return int(_lambda(n) + 1)

def chain_from_indvec(indvec):
    indvec_len = len(indvec)
    nums = [0] * (indvec_len + 1)
    nums[0] = 1

    for i in range(0, indvec_len):
        nums[i + 1] = nums[i] + nums[indvec[i] - 1]
    return nums


def indvec_gen(n, start = 1):
    indvec = [0] * n
    for i in range(0, n):
        indvec[i] = i + start
    return indvec

def indvec_is_min(indvec):
    for i in indvec:
        if i > 1:
            return False
    return True

def indvec_dec(indvec):
    indvec_len = len(indvec)

    for i in range(0, indvec_len):
        if indvec[indvec_len - 1 - i] > 1:
            indvec[len(indvec) - 1 - i] -= 1
            for j in range(indvec_len - i, indvec_len):
                indvec[j] = j + 1
            return

def indvec_get_last_num(indvec):
    indvec_len = len(indvec)
    nums = [0] * (indvec_len + 1)
    nums[0] = 1

    for i in range(0, indvec_len):
        nums[i + 1] = nums[i] + nums[indvec[i] - 1]
    return nums[indvec_len]


def indvec_crush(n):
    tries = [1, 1]

    lmin = math.ceil(math.log(n, 2))
    lmax = _lambda(n) + _nu(n) - 1

    for m in range(lmin, lmax + 1):
        q = int(m / 2)
        iv_fix = indvec_gen(q)

        while True:	# processing all possible fixed parts
            iv_mut = indvec_gen(m - q, q + 1)

            iv_fix.append(iv_mut[0])
            aq = indvec_get_last_num(iv_fix)
            del iv_fix[-1]
            amin = 1 + m - q          # a_{q+1} + m - q
            amax = aq * 2 ** (m - q)  # a_{q+1} * 2^{m-q}

            if n > amax or n < amin:  # n is not in [a_min, a_max], skipping this fixed part
                if indvec_is_min(iv_fix):
                    break
                indvec_dec(iv_fix)
                continue

            while True: # processing all possible mutable parts
                if indvec_get_last_num(iv_fix + iv_mut) == n:
                    return (iv_fix + iv_mut, tries)

                if indvec_is_min(iv_mut):
                    break
                indvec_dec(iv_mut)
                tries[1] += 1

            if indvec_is_min(iv_fix):
                break
            indvec_dec(iv_fix)
            tries[0] += 1


print("l*(2^n-1) <= l*(n) +  n - 1")
for n in range(2, 13):
    res = indvec_crush(n)
    resb = indvec_crush(2 ** n - 1)
    print("{:9} <= {:5} + {:2} - 1\n{}".format(len(resb[0]), len(res[0]), n, len(resb[0]) <= len(res[0]) + n - 1))

l*(2^n-1) <= l*(n) +  n - 1
        2 <=     1 +  2 - 1
True
        4 <=     2 +  3 - 1
True
        5 <=     2 +  4 - 1
True
        7 <=     3 +  5 - 1
True
        8 <=     3 +  6 - 1
True
       10 <=     4 +  7 - 1
True
       10 <=     3 +  8 - 1
True
       12 <=     4 +  9 - 1
True
       13 <=     4 + 10 - 1
True
       15 <=     5 + 11 - 1
True
       15 <=     4 + 12 - 1
True


Результаты проверки гипотезы Шольца-Брауэра для звёздных цепочек предоставлены в следующей таблице:

$l^*(2^n - 1)$| $l^*(n)$ | $n$ | $l^*(2^n - 1) \le l^*(n) + n - 1$
--- | --- | --- | ---
3 | 2 | 2 | +
5 | 3 | 3 | +
6 | 3 | 4 | +
8 | 4 | 5 | +
9 | 4 | 6 | +
11 | 5 | 7 | +
11 | 4 | 8 | +
13 | 5 | 9 | +
14 | 5 | 10 | +
16 | 6 | 11 | +
16 | 5 | 12 | +

Таким образом, было подтверждено, что гипотеза Шольца-Брауэра для звёздных цепочек верна для $2 \le n \le 12$ (для $1$ очевидно $1 \le 1 + 1 - 1$)

### Выводы
В ходе выполнения данной практической работы были сформированы представления об аддитивных цепочках, были выработаны умения реализовывать и применять алгоритмы для нахождения минимальных аддитивных цепочек, был сформирован навык использования системы компьютерной математики SageMath для реализации таких алгоритмов.

На практике была проверена гиптоеза Шольца-Брауэра, были проанализированы алгоритм Брауэра и алгоритм дробления векторов индексов.